packages/light.js/src/rpc/utils/createRpc.ts

```diff
@@ -3,99 +3,78 @@
-    if (typeof this.clear === 'function') {
-      this.clear();
-    }
-  }
-};
+import { createApiFromProvider, getApi } from '../../api';
+import { distinctValues } from '../../utils/operators';
+import { Metadata, RpcObservableOptions } from '../../types';
 
 /**
  * Add metadata to an RpcObservable, and transform it into a ReplaySubject(1).
+ * It's a currified function.
+ * Pure function version of {@link createRpc}.
  *
  * @ignore
  * @param metadata - The metadata to add.
- * @return - The original RpcObservable with patched metadata.
+ * @example
+ * createRpc(metadata)(options) returns a RpcObservable.
+ * createRpc(metadata)(options)(someArgs) returns an Observable.
  */
-const createRpc = <Source, Out>(metadata: Metadata<Source, Out>) => {
-  // rpc$ will hold the RpcObservable minus its metadata
-  const rpc$: RpcObservableWithoutMetadata<Source, Out> = (...args: any[]) => {
+const createRpcWithApi = memoizee(
+  <Source, Out>(metadata: Metadata<Source, Out>, api: any, ...args: any[]) => {
     // The source Observable can either be another RpcObservable (in the
     // `dependsOn` field), or anObservable built by merging all the
     // FrequencyObservables
     const source$ = metadata.dependsOn
-      ? metadata.dependsOn(...args)
-      : merge(...metadata.frequency);
-
-    // The last arguments is an options, if it's an object
-    // TODO What if we pass a single object as argument, which is not options?
-    const options: { withoutLoading?: boolean } =
-      args && args.length && isObject(args[args.length - 1]) ? args.pop() : {};
+      ? metadata.dependsOn(...args, { provider: api.provider })
+      : merge(...metadata.frequency.map(f => f({ provider: api.provider })));
 
-    // A RpcObservable is a source$ Observable, a single subject$ that
-    // subscribesthis source, and this subject$ multicasts the fired values to
-    // all Observers.
+    // A RpcObservable is: a source$ Observable, a single subject$ that
+    // subscribes to this source, and this subject$ multicasts the fired values
+    // to all Observers.
     const subject$ = new ReplaySubject<Out>(1);
 
     // The pipes to add
     const pipes: OperatorFunction<any, any>[] = [];
     if (metadata.pipes && isFunction(metadata.pipes)) {
-      pipes.push(...metadata.pipes(...args));
+      pipes.push(...metadata.pipes(api));
     }
-    pipes.push(multicast(() => subject$), refCount());
-    if (options.withoutLoading === true) {
-      pipes.push(withoutLoading());
+    pipes.push(multicast(() => subject$), refCount(), distinctValues());
+
+    return source$.pipe(...pipes) as Observable<Out>;
+  },
+  {
+    length: 3,
+    normalizer: (args: any) => {
+      // Custom memoization function, i.e. create an unique id from the args.
+      // `args` is arguments object as accessible in memoized function
+      return `${args[0].name}${args[1].provider.id}${JSON.stringify(
+        Array.from(args).slice(2)
+      )}`;
```

Comment:
This is the only place where I'm not so satisfied.

Normalizer is a function that given the args, generates a unique id used for memoization.
For args[0] = metadata, I take the name, that's alright.
For args[1] = api object, I use provider.id, but this is not guaranteed to be unique per provider (or is it?)
Then for the other args, I JSON.stringify them, that should be alright too.

---

I didn't find anything with `memoizee` to customize a compare function, where we'd use === for the api argument.

Reply:
Seems reasonable to me, I don't have any better idea

Reply:
https://github.com/medikoo/memoizee/issues/99